### PR TITLE
Hallucinations

### PR DESCRIPTION
**Pull request by FaDeOkno**
## About the PR

This PR adds real working hallucinations to the game.

## Why / Balance

Just, eh.. It's kinda strange to call SeeingRainbows effect as a hallucinations so I decided to make real ones.

## Technical details

So, basicly the system uses random VisibilityFlags so other people can not see hallucinations of others. That's pretty simple.
Also hallucinations entities spawns near humanoids, including the hallucinating person.
There also is a new prototype that contains details of hallucinations, so you can just make your ents and a pack that contains them to use.

## Media

https://github.com/space-wizards/space-station-14/assets/143940725/715c4b91-6a86-4eed-9bfb-3ae455ddc8f2

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: Added hallucinations.
- add: Now mindbreaker toxin causes hallucinations effect.

### CODE REVIEW

**Comment by metalgearsloth** [comment]
Needs a significant amount of reviewing as currently this will tank server performance.

**Comment by FaDeOkno** [comment]
I'd like to make it better, but not sure how yet

**Comment by IProduceWidgets** [comment]
Is there any reason for this to be doing anything on server except for ensuring the component?

**Comment by deltanedas** [comment]
client cant do ai

**Comment by FaDeOkno** [comment]
> Is there any reason for this to be doing anything on server except for ensuring the component? Seems like basically all of it should be on client since the stuff only displays to a single client.

Entities now are mobs so I don't think so
But maybe there is a way...

File: Content.Server/Hallucinations/HallucinationsSystem.cs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+            var uidhallucination = Spawn(_random.Pick(stat.Spawns), uidnewCoords);
+            EnsureComp<VisibilityComponent>(uidhallucination, out var uidvisibility);
+            _visibilitySystem.SetLayer(uidhallucination, uidvisibility, (int) stat.Layer, false);
+            _visibilitySystem.RefreshVisibility(uidhallucination, visibilityComponent: uidvisibility);
+
+        }
+    }
+
+}
```
Comment (metalgearsloth) [inline Content.Server/Hallucinations/HallucinationsSystem.cs:150]: All of this needs to be clientside. Adding another visibility layer which will have a significant impact on PVS is absolutely not worth it.

EntityManager is also a dependency on systems you do not need to get it again.

File: Content.Shared/Hallucinations/HallucinationsPrototype.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    [IdDataField]
+    public string ID { get; private set; } = default!;
+
+    /// <summary>
+    ///     List of prototypes that are spawned as a hallucination.
+    /// </summary>
+    [DataField("entities")]
+    public List<EntProtoId> Entities = new();
+
+    /// <summary>
```
Comment (metalgearsloth) [inline Content.Shared/Hallucinations/HallucinationsPrototype.cs:21]: Datafield strings are obsolete please remove them.

File: Content.Shared/Hallucinations/HallucinationsComponent.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    /// <summary>
+    /// How far from humanoid can appear hallucination
+    /// </summary>
+    [DataField, ViewVariables(VVAccess.ReadWrite)]
+    public float Range = 7f;
+
+    /// <summary>
+    /// How often (in seconds) hallucinations spawned
+    /// </summary>
+    [DataField, ViewVariables(VVAccess.ReadWrite)]
```
Comment (metalgearsloth) [inline Content.Shared/Hallucinations/HallucinationsComponent.cs:21]: This needs to be networked, TimeOffsetSerializer doesn't handle pausing on its own but we have a sourcegen version for that now, and datafield + VVRW is pointless.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Also needs file-scoped namespaces.

**Comment by bruhmogus** [comment]
this idea is funny as fuck ngl

**Pull request closed without merging**